reader_concurrency_semaphore.cc

```diff
@@ -1023,7 +1025,7 @@ void reader_concurrency_semaphore::consume(reader_permit::impl& permit, resource
 
 void reader_concurrency_semaphore::signal(const resources& r) noexcept {
     _resources += r;
-    maybe_admit_waiters();
+    _ready_list_cv.signal();
```

Comment:
I understand how this change improves some workloads, where a release has high chance of causing a new permit to be admitted. But can't it also regress? If a release doesn't cause a waiting request to be admitted, then we wasted a task wakeup for nothing.

Reply:
I also expected to see _read_list_cv.signal() removed from maybe_admit_waiters().

Reply:
Looking at maybe_admit_waiters(), I can see that moving it to execution_loop() makes sense. The two touch the same variables so one is heating the cache for the other. If they're separated by a wakeup, it goes away.

Example of a workload where this can be a regression: if the queue is usually empty. Then we waste a task to check that the queue is empty, and go back to sleep immediately.

Ideas:
1. retain maybe_admit_waiters() and make it do the minimum number of checks (perhaps none) and signal the condition variable; the old maybe_admit_waiters() can be called do_admit_waiters (and can contain the same checks, minus the conditions we moved to the new maybe_admit_waiters(), and minus the condition variable wakeup. (we have eager checks and lazy checks)
2. Consider moving the empty-queue check to the new maybe_admit_waiters()
3. Measure using perf-simple-query (it would usually have a non-empty queue so doesn't help with evaluating the empty-queue check)

My feeling is that the empty-queue check is worthwhile. Many workloads have enough resources to not queue. Checking early means we can yield to compaction or another service level or something.

Reply:
All single-partition reads are executed via the excution-loop, which means that all our latency sensitive reads already go through it. The loop does a poll after executing each read, so even if the poll initiated from `signal()` was deferred, at latest when the read is done, the loop polls again before going to sleep.
The exception is when the read defers, the poll might happen before the read released enough resources and the loop goes to sleep. This case wouldn't be helped with polling from `signal()`, as the read didn't release enough resources yet.
So I don't see a problem with deferring the poll.

Reply:
I'm talking about this sequence:

1. read queued
2. execution loop wakes executes read
3. read issues I/O
4. execution loop sleeps on empty queue
5. read completes, wakes execution loop
6. execution loop wakes, sees empty queue, sleeps again

This step 6 is wasteful, and not uncommon.

I'm proposing this:

1. read queued
2. execution loop wakes executes read
3. read issues I/O
4. execution loop sleeps on empty queue
5. read completes, sees queue is empty, does not wake execution loop

Basically moving the complex checks to the execution loop like the patch, but keeping the cheap check for an empty queue to avoid the wakeup.

It is very common for the queue to be empty, since the resources we reserve for reads are quite generous. It would be non-empty if we serve mostly from cache, or if serving very expensive reads like full scans.

Reply:
Ah, I see now. This makes sense, we can add yet another maybe_something:
```
void maybe_wake_execution_loop() {
  if (!_wait_queue.empty()) {
    _ready_list_cv.notify();
}
```
Then we call this instead of notifying the condition variable directly.

Reply:
> Yes, we can call it maybe_admit_waiters() and hten we don't need to change the call sites.

Yes, that works too, I'm fine with that.